okhttp/src/commonMain/kotlin/okhttp3/internal/-HostnamesCommon.kt

```diff
@@ -294,16 +296,40 @@ internal fun String.toCanonicalHost(): String? {
 
   val result = idnToAscii(host) ?: return null
   if (result.isEmpty()) return null
+  if (result.containsInvalidHostnameAsciiCodes()) return null
+  if (result.containsInvalidLabelLengths()) return null
 
-  return if (result.containsInvalidHostnameAsciiCodes()) {
-    // The IDN ToASCII result contains illegal characters.
-    null
-  } else if (result.containsInvalidLabelLengths()) {
-    // The IDN ToASCII result contains invalid labels.
-    null
-  } else {
-    result
+  return result
+}
+
+internal fun idnToAscii(host: String): String? {
+  val bufferA = Buffer().writeUtf8(host)
+  val bufferB = Buffer()
+
+  // 1. Map, from bufferA to bufferB.
+  while (!bufferA.exhausted()) {
+    val codePoint = bufferA.readUtf8CodePoint()
+    if(!IDNA_MAPPING_TABLE.map(codePoint, bufferB)) return null
   }
+
+  // 2. Normalize, from bufferB to bufferA.
+  val normalized = normalizeNfc(bufferB.readUtf8())
+  bufferA.writeUtf8(normalized)
+
+  // 3. For each label, convert/validate Punycode.
+  val decoded = Punycode.decode(bufferA.readUtf8()) ?: return null
+
+  // 4.1 Validate.
+
+  // Must be NFC.
+  if (decoded != normalizeNfc(decoded)) return null
+
+  // TODO: Must not begin with a combining mark.
```

Comment:
Will these TODOs correct the failing cases?

Reply:
Yes, but I’m not 100% certain we want to do these TODOs. Emitting the correct ASCII domain name for an input Unicode domain name is important and worth the effort. But to enforce things like joiners we need to haul in even more unicode data.